File: stack/ruby/stack.rb
require_relative 'node'

class Stack

    def initialize
        @top = nil
    end

    def push(data)
        new_node = Node.new(data)
        if @top != nil
            new_node.next_node = @top
        end

        @top = new_node
    end

    def pop 
        return nil if is_empty

        popped_node = @top
        @top = @top.next_node
        popped_node.data
    end

    def size
        count = 0
        current = @top 
        while current
            count += 1
            current = current.next_node
        end
        count
    end 

    def is_empty
        @top.nil?
    end

    def peek
        @top ? @top.data : nil
    end
end

File: stack/js/stack.js
class Node {
    constructor(data) {
        this.data = data;
        this.next = null;
    }
}

class Stack {
    constructor() {
        this.top = null
    }

    push(data) {
        const newNode = new Node(data);
        newNode.next = this.top;
        this.top = newNode;
    }

    pop() {
        if (this.top == null) {
            return null
        }

        const poppedNode = this.top;
        this.top = this.top.next;
        return poppedNode.data;
    }

    printStack() {
        let current = this.top;
        let result = []
        while (current) {
            result.push(current.data);
            console.log(current.data);
            current = current.next;
        }

        return result;
    }
}

module.exports = Stack;